ui/ops/src/dynamic/widgets/usage/usage.js

```diff
@@ -0,0 +1,192 @@
+import {computed, reactive, toValue} from 'vue';
+
+
+/**
+ * Returns the usage count for periods between edges.
+ * The usage at index i will be the total usage for the period between edges[i] and edges[i+1].
+ *
+ * @param {import('vue').MaybeRefOrGetter<string>} name
+ * @param {import('vue').MaybeRefOrGetter<Date[]>} edges
+ * @return {import('vue').ComputedRef<{x: Date, groups: Record<string, {x: Date, y: number, last: number}>}[]>}
+ */
+export function useUsageCount(name, edges) {
+  // As there's no way to aggregate results from the history api we have to do it ourselves.
+  // There could be quite a lot of data, we need to be careful not to keep it all around in memory.
+  // Because we only care about the max, we can process each pair of edges and just remember the max.
+
+  const countsByEdge = reactive(
+      /** @type {{number: Record<string, {x: Date, y: number, last: number}>}} */
+      {} // keyed by the leading edges .getTime()
+  );
+  asyncWatch([() => toValue(name), () => toValue(edges)], async ([name, edges], [oldName]) => {
+    if (name !== oldName) {
+      Object.keys(countsByEdge).forEach(k => delete countsByEdge[k]);
+    }
+
+    const toDelete = new Set(Object.keys(countsByEdge));
+    for (const edge of edges) {
+      toDelete.delete(edge.getTime().toString());
+    }
+    for (const k of toDelete) {
+      delete countsByEdge[k];
+    }
+
+    for (let i = 0; i < edges.length - 1; i++) {
+      const leadingEdge = edges[i];
+      if (!countsByEdge[leadingEdge.getTime()]) {
+        const toFetch = [leadingEdge];
+        for (let j = i + 1; j < edges.length; j++) {
+          const e = edges[j];
+          toFetch.push(e);
+          if (countsByEdge[e.getTime()]) {
+            i = j - 1; // will i++ in the loop
+            break;
+          }
+        }
+
+        const countBefore = countsByEdge[edges[i - 1]?.getTime()]?.last ?? null;
+        const records = await readUsageCountSeries(name, toFetch, countBefore);
+        for (const dataset in records) {
+          const groupData = records[dataset];
+          for (const record of groupData) {
+            if (!countsByEdge[record.x.getTime()])
+              countsByEdge[record.x.getTime()] = {};
+
+            countsByEdge[record.x.getTime()][dataset] = record.y;
+          }
+        }
+      }
+    }
+  }, {immediate: true});
+
+  return computed(() => {
+    return Object.entries(countsByEdge)
+        .map(([timestamp, groupData]) => ({
+          x: new Date(Number(timestamp)),
+          groups: groupData
+        }))
+        .sort((a, b) => a.x.getTime() - b.x.getTime());
+  });
+}
+
+/**
+ * Reads and calculates the total usage count for each span between the edges, returning it as a data series.
+ * The y property at index i in the response will be the total usage count for the period between edges[i] and edges[i+1].
+ * The last property at index i will be the most recent recorded usage count for the period between edges[i] and edges[i+1].
```

Comment:
There's no mention here about group IDs. I had to read the code to understand that this groups allocations by group ID before returning them.